tests/datasets/test_datasets.py

```diff
@@ -0,0 +1,79 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2020 Data61, CSIRO
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pytest
+import shutil
+import os
+from stellargraph.datasets import Cora, CiteSeer
+from urllib.error import URLError
+from stellargraph.datasets.dataset_loader import DatasetLoader
+from urllib.request import urlretrieve
+from unittest.mock import patch
+
+
+# use parametrize to automatically test each of the datasets that (directly) derive from DatasetLoader
+@pytest.mark.parametrize("dataset_class", list(DatasetLoader.__subclasses__()))
+def test_dataset_download(dataset_class):
+    dataset_class().download(ignore_cache=True)
+
+
+def test_invalid_url() -> None:
+    dataset = Cora()
+    dataset.url = "http://stellargraph-invalid-url/x"
+    with pytest.raises(URLError):
+        dataset.download(ignore_cache=True)
+
+
+def test_missing_files() -> None:
+    # we start with Cora to a special path, but with the URL from CiteSeer
+    dataset = Cora()
+    dataset.directory_name = "test-missing-files"
```

Comment:
Is this going to attempt to download to `~/stellargraph-datasets/test-missing-files`? Doesn't this not work as exactly intended because the download code relies on the directory matching the one in the archive, and this special directory doesn't?

Also, this could cause conflicts with concurrent test runs? I guess other tests (like the caching one) could fail in this case too so maybe we don't care?

Reply:
well spotted, I've replaced this with a much simpler test that doesn't need a special folder